### DISCUSSION TITLE
Syntactic analogy between components and tags.

### DISCUSSION THREAD

**Discussion by quimt** [HapticX/happyx#63] [Ideas]
It would be helpful in general if components placement in the DSL worked as if the component were an extended tag.

For instance, components might be some kind of complicated flex-container, or a fancy button with some of its styling dependent on the state of some `remember`ed vars.

Then might want to say
```
component FancyButton:
    `template`: tButton

appRoutes(...)
  "/":
        component FancyButton:
          "fancyText"

```
In analogy with 

```
Approutes(...)
  "/":
        tButton:
          "unfancyText"
```

to set the text child of the button component. Currently it does not work this way: the "fancyText" does not bind as child of the button. I predict this logic would be hard to code in a fully automated way since there are a couple different cases. Karax handles this in part with a distinction between `VNode`s, which are the actual elements of the VDOM, and `VComponent`s, which render by producing a `VNode` which renders the exact same way as other literal `VNode`s.

But perhaps the template macro in this framework could be extended to explicitly declare where children should attach when the component is placed this way in some other template or route. Maybe something in line with how the nim` {.importjs .} ` [pragma works.](https://nim-lang.org/docs/manual.html#implementation-specific-pragmas-importjs-pragma). For example, maybe:

```
component FancyButton:
  fancyParam: complexType = initComplexType(2)
    `template`: tButton:
       parse(@@1).recode.toText()
       @@rest
```
to apply some kind of fancy processing to the first child, but leave the @click part that comes after untouched.

**Comment by Ethosa** [comment]
> to set the text child of the button component. Currently it does not work this way: the "fancyText" does not bind as child of the button. I predict this logic would be hard to code in a fully automated way since there are a couple different cases

```nim
component FancyButton:
  `template`:
    tButton:
      slot

...

builtHtml:
  component FancyButton:
    "hello"
  component FancyButton:
    "world"
```

**Comment by Ethosa** [comment]
slots allow this. You can also use it like this
```nim
component Button:
  flat: bool = true
  `template`:
    if self.flat:
      tButton:
        slot
    else:
     tButton(...):
       slot
```

**Reply by quimt** [reply]
Ah, thanks for clarifying about the slots! At a glance I had thought they were part of some other declaration.
